### ISSUE TITLE
v0.31.1: palette-based RGBA PNG output can add transparent pixels

### ISSUE THREAD

**Issue by imranbarbhuiya** [lovell/sharp#3395]
## Possible bug
After upgrading to `0.31.1`, compress is changing image color. We are sending a png buffer to sharp to change it's quality to 95% and it was working fine. But now it's changing adding these black color
Image attached below
### Is this a possible bug in a feature of sharp, unrelated to installation?
yes

- [x] Running `npm install sharp` completes without error.
- [x] Running `node -e "require('sharp')"` completes without error.

If you cannot confirm both of these, please open an [installation issue](https://github.com/lovell/sharp/issues/new?labels=installation&template=installation.md) instead.

### Are you using the latest version of sharp?

- [x] I am using the latest version of `sharp` as reported by `npm view sharp dist-tags.latest`.

If you cannot confirm this, please upgrade to the latest version and try again before opening an issue.

If you are using another package which depends on a version of `sharp` that is not the latest, please open an issue against that package instead.

### What is the output of running `npx envinfo --binaries --system --npmPackages=sharp --npmGlobalPackages=sharp`?

```sh
  System:
    OS: Linux 5.15 Ubuntu 20.04.4 LTS (Focal Fossa)
    CPU: (24) x64 12th Gen Intel(R) Core(TM) i9-12900K
    Memory: 88.28 GB / 125.61 GB
    Container: Yes
    Shell: 5.0.17 - /bin/bash
  Binaries:
    Node: 16.16.0 - ~/.volta/tools/image/node/16.16.0/bin/node
    Yarn: 3.2.3 - ~/.volta/tools/image/yarn/1.22.19/bin/yarn
    npm: 8.11.0 - ~/.volta/tools/image/node/16.16.0/bin/npm
  npmPackages:
    sharp: ^0.31.1 => 0.31.1
```
### What are the steps to reproduce?

provide a png to sharp
```ts
const finalImage = await sharp(renderedImage).png({ quality: 94 }).toBuffer();
```
### What is the expected behaviour?
It should work like before without adding those black colors

### Please provide a minimal, standalone code sample, without other dependencies, that demonstrates this problem
```ts
// canvas is @napi-rs/canvas and it's output is working fine. The issue occurs only when passed to shard.
const renderedImage = canvas.png();
const finalImage = await sharp(renderedImage).png({ quality: 94 }).toBuffer();
```

### Please provide sample image(s) that help explain this problem
![image](https://cdn.discordapp.com/attachments/870354581115256852/1025751450070036520/a2ff75e5-f6a8-423c-84ee-551a9e330792.png)

**Comment by lovell** [comment]
> After upgrading to 0.31.1

From which version are you upgrading?

The sample image provided looks like a palette-based PNG. Is this the output? If so, please can you provide the input image.

**Comment by imranbarbhuiya** [comment]
From which version are you upgrading?
0.30.1
reverting to 0.30.0 works
<details>
<summary>Input:</summary>

![input](https://user-images.githubusercontent.com/74945038/193424132-b46a37b0-40be-44f3-bb46-d2042e233418.png)

</details>
<details>
<summary>Output:</summary>

![output](https://user-images.githubusercontent.com/74945038/193424167-fb95c03d-7418-497a-99d6-d47877b77348.png)

</details>

**Comment by lovell** [comment]
Thank you, an upstream fix for this is at https://github.com/libvips/libvips/pull/3074

**Comment by lovell** [comment]
sharp v0.31.2 now available with libvips v8.13.3

**Comment by imranbarbhuiya** [comment]
Thanks for the fix ❤️
